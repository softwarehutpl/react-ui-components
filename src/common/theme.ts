const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
