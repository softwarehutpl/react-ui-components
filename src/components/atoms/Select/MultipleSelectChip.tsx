import React from 'react';
import styled from 'styled-components';
import { CloseOutline } from '@styled-icons/evaicons-outline/CloseOutline';
import { defaultProps, IOption } from './Select';

interface IMultipleSelectChip {
  onDelete: () => void;
  option: IOption;
  color?: string;
  className?: string;
}

const MultipleSelectChip = ({ onDelete, option, className }: IMultipleSelectChip) => {
  return (
    <span className={className}>
      {option.label}
      <CloseOutline onClick={onDelete} style={{ width: '12px' }} />
    </span>
  );
};

export default styled(MultipleSelectChip)<IMultipleSelectChip>`
  ${({ theme, color = defaultProps.color }) => ({
    padding: '5px',
    backgroundColor: theme.colors[color].light,
    color: theme.colors[color].base,
    'margin-right': '5px',
    'font-size': '0.75em',
  })}
`;
