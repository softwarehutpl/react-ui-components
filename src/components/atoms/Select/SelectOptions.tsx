import styled from 'styled-components';

export interface ISelectOptions {
  color?: string;
  toggleHeight: number;
}

export const SelectOptions = styled.div<ISelectOptions>`
  ${({ toggleHeight }) => ({
    width: '100%',
    position: 'absolute',
    left: 0,
    top: `${toggleHeight}px`,
    'box-shadow': '0 4px 10px 4px rgba(217,217,217,0.6)',
    'max-height': '200px',
    'overflow-y': 'scroll',
  })}
`;
