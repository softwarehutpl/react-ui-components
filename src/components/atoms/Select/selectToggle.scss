.multipleOptions {
  max-width: 80%;
  text-align: left;
}
