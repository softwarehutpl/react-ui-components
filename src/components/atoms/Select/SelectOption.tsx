import React from 'react';
import styled from 'styled-components';
import { defaultProps } from './Select';

interface ISelectOptionProps {
  selectOnChange?: (option: { label: string, value: string | number }) => void;
  className?: string;
  option: { label: string, value: string | number };
  padding?: number;
  color?: string;
  optionsBackgroundColor?: string;
  optionsFontColor?: string;
}

const SelectOption = ({
  selectOnChange,
  className,
  option,
}: ISelectOptionProps) => {
  const handleClick = () => {
    if (selectOnChange) {
      selectOnChange(option);
    }
  };
  return (
    <div className={className} onClick={handleClick}>
      {option.label}
    </div>
  );
};

const StyledSelectOption = styled(SelectOption)<ISelectOptionProps>`
  ${({ theme, color = defaultProps.color, padding = defaultProps.padding, optionsBackgroundColor, optionsFontColor }) => ({
    width: '100%',
    padding: `${padding / 2}px ${padding}px`,
    'background-color': optionsBackgroundColor || theme.colors[color].light,
    color: optionsFontColor || theme.colors[color].base,
    '&:hover': {
      'background-color': theme.colors[color].dark,
    }
  })}
`;

export default StyledSelectOption;
