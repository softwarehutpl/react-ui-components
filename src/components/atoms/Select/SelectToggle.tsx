import styled from 'styled-components';
import { defaultProps } from './Select';

export interface ISelectToggle {
  color?: string;
  backgroundColor?: string;
  fontColor?: string;
  padding?: number;
}

export const SelectToggle = styled.div<ISelectToggle>`
  ${({ theme, color = defaultProps.color, backgroundColor, fontColor, padding }) => ({
    'background-color': backgroundColor || theme.colors[color].base,
    color: fontColor || theme.colors[color].light,
    width: '100%',
    padding: `${padding}px`,
    display: 'flex',
    'justify-content': 'space-between',
  })}
`;
