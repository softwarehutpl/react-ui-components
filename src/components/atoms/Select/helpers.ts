import { IOption } from './Select';

const isEqual = require('lodash.isequal');

export const checkIfIsSelected = (
  option: IOption,
  multiple?: boolean,
  value?: IOption,
  multipleValue?: IOption[]
) => {
  if (multiple && multipleValue) {
    return !!multipleValue.find((val) => isEqual(val, option));
  }
  return isEqual(value, option);
};
