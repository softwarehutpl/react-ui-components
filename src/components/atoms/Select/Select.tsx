import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import useOutsideClick from '../../../common/hooks/clickOutside';
import SelectToggle from './SelectToggle';
import { SelectOptions } from './SelectOptions';
import { SelectOption } from './SelectOption';
import { checkIfIsSelected } from './helpers';

export interface IOption {
  label: string;
  value: string | number;
}

interface ISelect {
  value?: IOption;
  multipleValue?: IOption[];
  onChange: (newValue: IOption | IOption[]) => void;
  options: IOption[];
  placeholder?: string;
  multiple?: boolean;
  color?: string;
  className?: string;
  disabled?: boolean;
  backgroundColor?: string;
  fontColor?: string;
  padding?: number;
  caretSize?: number;
  width?: number;
  optionsBackgroundColor?: string;
  optionsFontColor?: string;
  selectedOptionBackgroundColor?: string;
}

interface IStyledSelect {
  margin?: number;
  width?: number;
}

const Select = ({
  options,
  color,
  className,
  value,
  multipleValue,
  onChange,
  placeholder,
  backgroundColor,
  fontColor,
  padding,
  optionsBackgroundColor,
  optionsFontColor,
  multiple,
  caretSize,
  selectedOptionBackgroundColor,
  disabled,
}: ISelect) => {
  const [isOpen, setIsOpen] = useState(false);
  const [toggleHeight, setToggleHeight] = useState(0);
  const toggleRef = React.useRef<HTMLDivElement>(null);
  const selectRef = React.useRef<HTMLDivElement>(null);

  const toggleSelect = () => {
    if (!disabled) {
      setIsOpen(!isOpen);
    }
  };

  const closeSelect = () => {
    setIsOpen(false);
  };

  const countToggleHeigh = () => {
    if (toggleRef && toggleRef.current) {
      const { height } = toggleRef.current.getBoundingClientRect();
      setToggleHeight(height);
    }
  };

  useOutsideClick(selectRef, closeSelect);

  useEffect(() => {
    countToggleHeigh();
  }, []);

  const singleSelectOnChange = (option: IOption) => {
    onChange(option);
    closeSelect();
  };

  const multipleSelectOnChange = (option: IOption, selectedOptions?: IOption[]) => {
    const newSelectedOptions = selectedOptions ? [...selectedOptions] : [];
    if (!newSelectedOptions.includes(option)) {
      newSelectedOptions.push(option);
      onChange(newSelectedOptions);
      setTimeout(countToggleHeigh, 100);
    }
  };

  const onMultipleOptionDelete = (option: IOption, selectedOptions?: IOption[]) => {
    if (selectedOptions) {
      const newOptions = selectedOptions.filter((opt) => opt !== option);
      onChange(newOptions);
      setTimeout(countToggleHeigh, 100);
    }
  };

  return (
    <div className={className} ref={selectRef}>
      <SelectToggle
        ref={toggleRef}
        toggle={toggleSelect}
        color={color}
        backgroundColor={backgroundColor}
        fontColor={fontColor}
        padding={padding}
        disabled={disabled}
        onChange={(newOptions: IOption[]) => {
          onChange(newOptions);
        }}
        onMultipleOptionDelete={(option: IOption) => {
          onMultipleOptionDelete(option, multipleValue);
        }}
        value={value}
        placeholder={placeholder}
        multipleValue={multipleValue}
        caretSize={caretSize}
      />
      {isOpen && (
        <SelectOptions toggleHeight={toggleHeight} color={color}>
          {options.map((option) => (
            <SelectOption
              key={option.label}
              padding={padding}
              color={color}
              optionsFontColor={optionsFontColor}
              optionsBackgroundColor={optionsBackgroundColor}
              onClick={() => {
                if (multiple) {
                  multipleSelectOnChange(option, multipleValue);
                } else {
                  singleSelectOnChange(option);
                }
              }}
              isSelected={checkIfIsSelected(option, multiple, value, multipleValue)}
              selectedOptionBackgroundColor={selectedOptionBackgroundColor}
            >
              {option.label}
            </SelectOption>
          ))}
        </SelectOptions>
      )}
    </div>
  );
};

export const defaultProps = {
  color: 'primary',
  width: 200,
  margin: 0,
  placeholder: 'select value',
  padding: 15,
  multiple: false,
  caretSize: 15,
  disabled: false,
};

Select.defaultProps = defaultProps;

const StyledSelect = styled(Select)<IStyledSelect>`
  ${({ margin = defaultProps.margin, width = defaultProps.width }) => ({
    margin: `${margin}px`,
    position: 'relative',
    width: `${width}px`,
  })}
`;

export default StyledSelect;
