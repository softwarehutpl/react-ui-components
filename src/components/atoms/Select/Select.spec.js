import React from 'react';
import { create, act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import theme from '../../../common/theme';
import Select from './Select';
import {SelectToggle} from './SelectToggle';
import SelectOption from './SelectOption';

describe('Select component', () => {
  it('should render select toggle with given placeholder', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select placeholder="foo" options={[]} />
      </ThemeProvider>
    );
    const instance = component.root;
    const selectToggle = instance.findByType(SelectToggle);
    expect(selectToggle.props.children[0]).toBe('foo');
  });

  it('should show select given options after click', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[
            { label: 'one', value: '1' },
            { label: 'two', value: '2' },
          ]}
        />
      </ThemeProvider>
    );
    const instance = component.root;
    const toggle = instance.findByType(SelectToggle);
    act(() => {
      toggle.props.onClick();
    })
    const options = instance.findAllByType(SelectOption);
    const optionOne = options[0].findByType('div');
    const optionTwo = options[1].findByType('div');
    expect(optionOne.props.children).toBe('one');
    expect(optionTwo.props.children).toBe('two');
  });

  it('should call onChange method', () => {
    const mockOnChange = jest.fn();
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[
            { label: 'one', value: 1 },
          ]}
          onChange={mockOnChange}
        />
      </ThemeProvider>
    );
    const instance = component.root;
    const toggle = instance.findByType(SelectToggle);
    act(() => {
      toggle.props.onClick();
    })
    const selectOption = instance.findByType(SelectOption);
    const option = selectOption.findByType('div');
    act(() => {
      option.props.onClick();
    })
    expect(mockOnChange.mock.calls.length).toEqual(1);
  });

  it('should render select with default style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[]}
        />
      </ThemeProvider>
    );
    const select = component.toJSON();
    expect(select).toHaveStyleRule('width', '200px');
    expect(select).toHaveStyleRule('margin', '0px');
  });

  it('should render select with custom style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[]}
          width={400}
          margin={10}
        />
      </ThemeProvider>
    );
    const select = component.toJSON();
    expect(select).toHaveStyleRule('width', '400px');
    expect(select).toHaveStyleRule('margin', '10px');
  });

  it('should render select toggle with default style for default primary color', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[]}
        />
      </ThemeProvider>
    );
    const select = component.toJSON();
    const selectToggle = select.children[0];
    expect(selectToggle).toHaveStyleRule('background-color', '#000000');
    expect(selectToggle).toHaveStyleRule('color', '#ffffff');
    expect(selectToggle).toHaveStyleRule('padding', '15px');
  })

  it('should render select toggle with custom style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[]}
          fontColor="red"
          backgroundColor="blue"
          padding={10}
        />
      </ThemeProvider>
    );
    const select = component.toJSON();
    const selectToggle = select.children[0];
    expect(selectToggle).toHaveStyleRule('background-color', 'blue');
    expect(selectToggle).toHaveStyleRule('color', 'red');
    expect(selectToggle).toHaveStyleRule('padding', '10px');
  })

  it('should render select option with default style for default primary color', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[
            { label: 'one', value: 1 },
          ]}
        />
      </ThemeProvider>
    );
    const instance = component.root;
    const toggle = instance.findByType(SelectToggle);
    act(() => {
      toggle.props.onClick();
    })
    const option = component.toJSON().children[1].children[0];
    expect(option).toHaveStyleRule('background-color', '#ffffff');
    expect(option).toHaveStyleRule('color', '#000000');
    expect(option).toHaveStyleRule('padding', '7.5px 15px');
  })

  it('should render select option with custom style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Select
          options={[
            { label: 'one', value: 1 },
          ]}
          optionsBackgroundColor="red"
          optionsFontColor="blue"
          padding={20}
        />
      </ThemeProvider>
    );
    const instance = component.root;
    const toggle = instance.findByType(SelectToggle);
    act(() => {
      toggle.props.onClick();
    })
    const option = component.toJSON().children[1].children[0];
    expect(option).toHaveStyleRule('background-color', 'red');
    expect(option).toHaveStyleRule('color', 'blue');
    expect(option).toHaveStyleRule('padding', '10px 20px');
  })
});
