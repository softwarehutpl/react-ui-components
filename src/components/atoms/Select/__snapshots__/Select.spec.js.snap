// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render correctly 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  background-color: #000000;
  color: #ffffff;
  width: 100%;
  padding: 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  box-sizing: border-box;
}

.c0 {
  margin: 0px;
  position: relative;
  width: 200px;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    foo
    <svg
      aria-hidden="true"
      className="c2"
      fill="currentColor"
      focusable="false"
      style={
        Object {
          "width": "15px",
        }
      }
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.178 19.569a.998.998 0 001.644 0l9-13A.999.999 0 0021 5H3a1.002 1.002 0 00-.822 1.569l9 13z"
      />
    </svg>
  </div>
</div>
`;
